### DISCUSSION TITLE
Mocking backend responses

### DISCUSSION THREAD

**Discussion by meirka** [testing-library/angular-testing-library#279] [Q&A]
Hello, is there a way to mock backend responses?

In case of TestBed, when I have a component with service that calls backend, I use HttpTestingController to respond to a request. How do I go about it with testing library?

**Comment by timdeschryver** [comment]
You can mock the service - see the examples https://github.com/testing-library/angular-testing-library/blob/main/apps/example-app/src/app/examples/12-service-component.spec.ts.

Another option is to use MSWjs - https://timdeschryver.dev/blog/using-msw-in-an-angular-project

**Reply by meirka** [reply]
https://github.com/testing-library/angular-testing-library/blob/main/apps/example-app/src/app/examples/12-service-component.spec.ts - mocks entire service, I would like to just mock backend.

Thank you.

**Comment by marcioggs** [comment]
You can mock `HttpClient` using a spy object.

e.g. using Jasmine:
```ts
  import { HttpClient } from '@angular/common/http';

  const httpClient: jasmine.SpyObj<HttpClient> = jasmine.createSpyObj('HttpClient', ['post']);
  httpClient.post.and.returnValue('Mocked response');

  // TODO: Provide httpClient on Angular Testing Library providers array.
  // { provide: HttpClient, useValue: httpClient }
```

**Comment by flensrocker** [comment]
You can use the `injector` on the `debugElement` of the object, the `render` function returns.

```typescript
const renderResult = await render(YourComponent, {
  providers: [provideHttpClient(), provideHttpClientTesting()],
})

const http = renderResult.debugElement.injector.get(HttpTestingController);
```
